[wallet] Verify unconfirmed transactions after syncing

afilini opened the pull request:
### Description

Verify the unconfirmed transactions we download against the consensus rules. This is currently exposed as an extra `verify` feature, since it depends on a pre-release version of `bitcoinconsensus`.

Closes #352

### Notes to the reviewers

Still missing some docs items but the code should be pretty much done

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

afilini committed:
[wallet] Verify unconfirmed transactions after syncing

Verify the unconfirmed transactions we download against the consensus
rules. This is currently exposed as an extra `verify` feature, since it
depends on a pre-release version of `bitcoinconsensus`.

Closes #352

RCasatta commented on src/wallet/verify.rs:
```diff
@@ -0,0 +1,171 @@
+    for (index, input) in tx.input.iter().enumerate() {
+        let prev_tx = if let Some(prev_tx) = database.get_raw_tx(&input.previous_output.txid)? {
+            prev_tx
+        } else if let Some(prev_tx) = blockchain.get_tx(&input.previous_output.txid)? {
```
we may want to locally cache the tx returned here to improve performance in a situation where txs is spending more input from the same tx. Maybe too edge case...

RCasatta commented on src/wallet/verify.rs:
```diff
@@ -0,0 +1,171 @@
+
+impl std::error::Error for VerifyError {}
+
+impl From<Error> for VerifyError {
```
We have `impl_error!` for this and the following

afilini replied:
I don't think I can use it for `From<crate::Error>` because we have to wrap that in a `Box`, but I can use it for the other one

RCasatta commented on src/wallet/verify.rs:
```diff
@@ -0,0 +1,171 @@
+
+    #[test]
+    fn test_verify_fail_unsigned_tx() {
+        let prev_tx: Transaction = deserialize(&Vec::<u8>::from_hex("020000000101192dea5e66d444380e106f8e53acb171703f00d43fb6b3ae88ca5644bdb7e1000000006b48304502210098328d026ce138411f957966c1cf7f7597ccbb170f5d5655ee3e9f47b18f6999022017c3526fc9147830e1340e04934476a3d1521af5b4de4e98baf49ec4c072079e01210276f847f77ec8dd66d78affd3c318a0ed26d89dab33fa143333c207402fcec352feffffff023d0ac203000000001976a9 ...
```
if this txs are on mainet or testnet maybe add a comment with the txid so that one could look on block explorer

RCasatta commented on src/wallet/verify.rs:
```diff
@@ -0,0 +1,171 @@
+
+    Consensus(bitcoinconsensus::Error),
+
+    Global(Box<Error>),
```
Maybe a comment why the Box is needed since Error is an enum (because Error contains a reference to VerifyError that would create a cycle)

afilini committed:
[verify] Cache txs to avoid multiple db/network lookups

afilini committed:
[verify] Add documentation

afilini committed:
[verify] Use `impl_error!()` whenever possible

afilini committed:
[verify] Build the `verify` feature on docs.rs

afilini committed:
[wallet] Verify unconfirmed transactions after syncing

Verify the unconfirmed transactions we download against the consensus
rules. This is currently exposed as an extra `verify` feature, since it
depends on a pre-release version of `bitcoinconsensus`.

Closes #352
